File: problems/problems_676/solution.ts
class TrieNode {
	public children: Map<string, TrieNode>;
	public isEnd: boolean;
	constructor() {
		this.children = new Map<string, TrieNode>();
		this.isEnd = false;
	}
	insert(word: string): void {
		let node: TrieNode = this;
		for (let i: number = 0; i < word.length; i++) {
			const ch: string = word[i];
			if (!node.children.has(ch)) {
				node.children.set(ch, new TrieNode());
			}
			node = node.children.get(ch);
		}
		node.isEnd = true;
	}
}

class MagicDictionary {
	private root: TrieNode;
    constructor() {
        this.root = new TrieNode();
    }

    buildDict(dictionary: string[]): void {
        for (const word of dictionary) {
			this.root.insert(word);
		}
    }

    search(searchWord: string): boolean {
        const query: Function = (node: TrieNode, index: number, remain: number): boolean => {
			if (index == searchWord.length) {
				return node.isEnd && remain == 0;
			}
			if (node.children.has(searchWord[index])) {
				if (query(node.children.get(searchWord[index]), index + 1, remain)) {
					return true;
				}
			}
			if (remain-- == 0) {
				return false;
			}
			// @ts-ignore
			for (const [k, v] of node.children.entries()) {
				if (k === searchWord[index]) {
					continue;
				}
				if (query(v, index + 1, remain)) {
					return true;
				}
			}
			return false;
		}
		return query(this.root, 0, 1);
    }
}

/**
 * Your MagicDictionary object will be instantiated and called as such:
 * var obj = new MagicDictionary()
 * obj.buildDict(dictionary)
 * var param_2 = obj.search(searchWord)
 */

export function Solve(inputJsonElement: string): any {
	const inputValues: string[] = inputJsonElement.split("\n");
	const operators: string[] = JSON.parse(inputValues[0]);
	const opValues: any[][] = JSON.parse(inputValues[1]);
	const ans: any[] = [null];
	const obj: MagicDictionary = new MagicDictionary();
	for (let i: number = 1; i < operators.length; i++) {
		if (operators[i] == "buildDict") {
			obj.buildDict(opValues[i][0]);
			ans.push(null);
			continue;
		}
		if (operators[i] == "search") {
			ans.push(obj.search(opValues[i][0]));
			continue;
		}
		ans.push(null);
	}
	return ans;
}

File: problems/problems_676/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"


using namespace std;
using json = nlohmann::json;

class TrieNode {
public:
    TrieNode* children[26];
    bool isEnd;
    TrieNode() {
        for (int i = 0; i < 26; i++) {
            children[i] = nullptr;
        }
        isEnd = false;
    }

    void insert(string word) {
        TrieNode* node = this;
        for (char c : word) {
            if (node->children[c - 'a'] == nullptr) {
                node->children[c - 'a'] = new TrieNode();
            }
            node = node->children[c - 'a'];
        }
        node->isEnd = true;
    }
};

class MagicDictionary {
private:
    TrieNode* root;
    bool query(TrieNode* node, string word, size_t idx, int remain) {
        if (idx == word.size()) {
            return remain == 0 && node->isEnd;
        }
        int cur = word[idx] - 'a';
        if (node->children[cur] != nullptr) {
            if (query(node->children[cur], word, idx + 1, remain)) {
                return true;
            }
        }
        if (remain-- == 0) {
            return false;
        }
        for (int i = 0; i < 26; i++) {
            if (i == cur || node->children[i] == nullptr) {
                continue;
            }
            if (query(node->children[i], word, idx + 1, remain)) {
                return true;
            }
        }
        return false;
    }
public:
    MagicDictionary() {
        root = new TrieNode();
    }
    
    void buildDict(vector<string> dictionary) {
        for (string word : dictionary) {
            root->insert(word);
        }
    }
    
    bool search(string searchWord) {
        return query(root, searchWord, 0, 1);
    }
};

/**
 * Your MagicDictionary object will be instantiated and called as such:
 * MagicDictionary* obj = new MagicDictionary();
 * obj->buildDict(dictionary);
 * bool param_2 = obj->search(searchWord);
 */

json leetcode::qubh::Solve(string input_json_values) {
	vector<string> inputArray;
	size_t pos = input_json_values.find('\n');
	while (pos != string::npos) {
		inputArray.push_back(input_json_values.substr(0, pos));
		input_json_values = input_json_values.substr(pos + 1);
		pos = input_json_values.find('\n');
	}
	inputArray.push_back(input_json_values);

	vector<string> operators = json::parse(inputArray[0]);
	vector<vector<json>> op_values = json::parse(inputArray[1]);
	auto obj0 = make_shared<MagicDictionary>();
	vector<json> ans = {nullptr};
	for (size_t i = 1; i < op_values.size(); i++) {
		if (operators[i] == "buildDict") {
			obj0->buildDict(op_values[i][0]);
			ans.push_back(nullptr);
			continue;
		}
		if (operators[i] == "search") {
			ans.push_back(obj0->search(op_values[i][0]));
			continue;
		}
		ans.push_back(nullptr);
	}
	return ans;
}
